File: nim/domemo.nim
#
# domemo.nim
#
# Copyright (c) 2018 Hiroaki Wada
#
# This software is released under the MIT License.
# http://opensource.org/licenses/mit-license.php
#
import random
import sequtils
import strutils
import algorithm

type
    Player = ref object of RootObj
        name*: string
        cards: seq[int]

proc check(player:Player): bool = return player.cards.len == 0

proc match(player:Player, guessNumber:int): int = 
    if player.cards.contains(guessNumber):
        player.cards.del(player.cards.find(guessNumber))
        player.cards.sort(system.cmp[int])
        echo "jackpot."
        return guessNumber
    else:
        echo "miss."
        return 0

proc show(player:Player): void =
    var cardStr = ""
    for card in player.cards:
        if player.name == "Player1":
            cardStr = cardStr & "X"
        else:
            cardStr = cardStr & $card
    echo player.name, " -> ", cardStr

proc show(cards:openarray[int], label:string): void =
    var cardStr = ""
    for card in cards:
        cardStr = cardStr & $card
    echo label, " -> ", cardStr

proc showCard(players:openarray[Player]): void =
    for player in players:
        player.show()

# Main Logic ---------------------------------------
randomize()

var cards:seq[int] = @[1,2,2,3,3,3,4,4,4,4,5,5,5,5,5,6,6,6,6,6,6,7,7,7,7,7,7,7]
cards.shuffle()

var openCards:seq[int] = cards[cards.len-4..cards.len-1]
openCards.sort(system.cmp[int])

var players:seq[Player] = @[]
for i in countup(0, 3):
    var playerCards:seq[int] = cards[i*5..(i+1)*5-1]
    playerCards.sort(system.cmp[int])
    players.add(Player(name: "Player" & $(i+1), cards: playerCards))

block mainLoop:
    while true:
        echo "---------"
        openCards.show("openCards")

        showCard(players)

        for player in players:
            var guessNumber:int = 0
            if player.name == "Player1":
                guessNumber = readLine(stdin).parseInt
            else:
                guessNumber = rand(7) + 1
        
            echo player.name, " is called -> ", guessNumber

            let result = player.match(guessNumber)
            if result > 0:
                openCards.add(result)
                openCards.sort(system.cmp[int])
            
            if player.check:
                echo player.name, " is Win!"
                break mainLoop

File: swift/domemo.swift
/*
  domemo.swift

  Copyright (c) 2018 Hiroaki Wada

  This software is released under the MIT License.
  http://opensource.org/licenses/mit-license.php
*/
import Foundation
import Glibc

class Player{
    var name: String
    var cards: [Int]

    init(name: String, cards: [Int]){
        self.name = name
        self.cards = cards
    }

    func show(){
        print(self.name + " -> ")
        showCards(cards: self.cards)
    }

    func check() -> Bool {
        return self.cards.count == 0
    }

    func match(guessCard: Int) -> Bool{
        if let idx = self.cards.index(of: guessCard){
            self.cards.remove(at: idx)
            print("jackpot!")
            return true
        }else{
            print("miss!")
            return false
        }
    }
}

func showCards(cards:[Int]){
    var str = ""
    for card in cards{
        str += "\(cards)"
    }
    print(str)
}

let cards:[Int] = [1,2,2,3,3,3,4,4,4,4,5,5,5,5,5,6,6,6,6,6,6,7,7,7,7,7,7,7]
for idx in cards.indices{
    let swp = cards[idx]
    let targetIdx = Int.random(in: 0 ..< 28)
    cards[idx] = cards[targetIdx]
    cards[targetIdx] = swp
}
var openCards:[Int] = Array(cards[cards.count ..< cards.count])

var players: [Player] = []
for idx in 0 ..< 4{
    players.append(Player(name: "Player\(idx)", cards: Array(cards[idx*5 ..< (idx+1)*5])))
}

var round = 1
MainLoop:while round < 100 {
    print("*round(\round)")
    var str = "openCard -> "
    showCards(cards: openCards)

    for player in players{
        player.show()
    }

    for player in players{
        let guessCard: Int = arc4random_uniform(7) + 1
        print("\(player.name) is Called -> \(guessCard)")
        if player.match(guessCard: guessCard){
            openCards.append(guessCard)
            if player.check(){
                print("\(player.name) is Win!")
                break MainLoop
            }
        }
    }
    round += 1
}